### ISSUE TITLE
Open Review enhancements in OJS/OMP

### ISSUE THREAD

**Issue by ajnyga** [pkp/pkp-lib#2976]
This is my plan to implement a better open review option to OJS/OMP. All comments are wellcome at this point @asmecher, @NateWr, @bozana and others.

# Additions to Reviewer workflow

A discussions grid visible in steps 3 and 4. The discussions visible are the same for each step.
In open reviews ability to start discussions both with the editors and the author. 
In double blind reviews ability to start discussions with the editor, see issue https://github.com/pkp/pkp-lib/issues/2717 and https://github.com/pkp/pkp-lib/issues/1257.

# Additions to Author workflow

In open reviews, the discussions grid enables the author to communicate with the reviewers for the open review requests.

The Reviewers grid is visible for the author and the author can see the review schedules of the review requests that were selected as open. When the open reviews become available, the author is notified and she can access them directly. If the same submission also has double blind review requests, these are not visible or available in discussions.

# Additions to Editor workflow

In open reviews, the editor can initiate discussions together with the author and the open request reviewers. 

# An Open Review workflow in action

1. The editor chooses the reviewers and selects open review for the requests
2. The reviewers accept the request. The author is notified about the reviewers and the Reviewer grid becomes visible in the author view and the author can initiate discussions with the reviewers.
3. In step 3 of the reviewer workflow, the reviewer can initiate discussions with the author if needed during the review.
4. When the reviews are send, the author is notified and the reviews are visible for the author
5. The reviewer (in step 4) and author can still initiate discussions if needed
6. The editor makes a normal editorial decision
7. Move on to copyediting or request changes or changes + second round

General questions:
- I think that it would make sense that the author finds out about the reviewers only after they have accepted the request. What do you think?
- Are there too many items in the review workflow step 3 if we add the discussions grid there?

**Comment by NateWr** [comment]
Great writeup! :+1: 

> A discussions grid visible in steps 3 and 4. The discussions visible are the same for each step.

Can this be the existing Review Discussions grid? It'd be good to centralize all discussions for the stage in one grid -- with reviewers/authors only seeing what's appropriate to them.

> When the reviews are send, the author is notified and the reviews are visible for the author

Could this just be a new discussion (automatically) opened with the author/reviewer/editor and the content of the review in the discussion?

> The editor makes a normal editorial decision

If we stored the IDs of the discussions that were created with each review in step 4, a new message could be posted to the discussion, which would effectively alert each reviewer/author of the decision.

In this way, each review discussion can act like a faceted log of activity, containing details about that review, but also being alerted to "global" events (like editorial decisions).

> I think that it would make sense that the author finds out about the reviewers only after they have accepted the request. What do you think?

:+1: 

> Are there too many items in the review workflow step 3 if we add the discussions grid there?

Not sure but that will be easier to workout down the line when we can see it in place.

Another consideration is how to surface these reviewer-facing discussions to reviewers in the submissions list.

**Comment by ajnyga** [comment]
Thanks @NateWr

> Can this be the existing Review Discussions grid? It'd be good to centralize all discussions for the stage in one grid -- with reviewers/authors only seeing what's appropriate to them.

Yes, I think we should definitely use the same review discussions grid in both those two steps. Meaning that the editors will also see all active discussions between the author and the reviewers in a specific review round.

> Could this just be a new discussion (automatically) opened with the author/reviewer/editor and the content of the review in the discussion?

Why not, but that would mean that the content of the review would be visible in two places: the discussions grid and the reviewer grid (link to the actual review). But I guess that is not a bad thing?

> If we stored the IDs of the discussions that were created with each review in step 4, a new message could be posted to the discussion, which would effectively alert each reviewer/author of the decision.
> In this way, each review discussion can act like a faceted log of activity, containing details about that review, but also being alerted to "global" events (like editorial decisions).

The author does already see the decision when it is made (a link appears in the top of the author view leading to a modal window). But are you thinking that in all review types we would move the editorial decisions to the discussions grid? Or that the decision would be visible in two places: in the discussions grid and the top of the page.

Currently, I think, the reviewers are not informed about editorial decisions. I am not sure if this is something that the editors would want? Or something that would just fit the scope of open reviews?

**Comment by NateWr** [comment]
> But are you thinking that in all review types we would move the editorial decisions to the discussions grid? Or that the decision would be visible in two places: in the discussions grid and the top of the page.

Two places. I was really thinking of it as just reporting back to the reviewer: here's a message that this decision was made. The actual decision record would still be where it is now.

> Currently, I think, the reviewers are not informed about editorial decisions. I am not sure if this is something that the editors would want? Or something that would just fit the scope of open reviews?

Yeah I don't know either. But I'm generally in favor of more feedback where possible. People like to see results from their efforts.

**Comment by ajnyga** [comment]
I actually have one journal doing open reviews. I could ask them, what they think, but ideally this would require more opinions. 

I guess this would fit the idea of an open review. Of course the discussion message with the decision would then only include the decision - not the reports of the other reviewers, I think?

But overall I think that the basics of the enhancements are fairly staightforward.

**Comment by NateWr** [comment]
> Of course the discussion message with the decision would then only include the decision - not the reports of the other reviewers, I think?

Mmm, yeah. It could complicate the workflow of taking the editorial action.

**Comment by ajnyga** [comment]
While discussing the new features with one of our journals they asked whether it would be possible to change the review type after the review is ready. The logic here is that the reviewer could agree to switch to open review more easily after seeing the whole text. It would also make the whole workflow more flexible.

This could be a nice feature in general. We would only add a new option to the edit review modal where you can already change the schedule of the review. What do you think @NateWr ?

I would like to hear comments from @stranack as well to the discussions and suggestions above.

**Comment by NateWr** [comment]
Are you suggesting the reviewer or the editor would be able to switch the review type?

**Comment by ajnyga** [comment]
The editor would be able to change the type any time after the request has been sent.

**Comment by NateWr** [comment]
Sounds ok to me as long as the reviewer is notified. Is that not possible now?

**Comment by ajnyga** [comment]
I dont think so. If you edit the review request, you can only change the dates and the review files selections. I think it would be fairly straight forward to add the new selection there.

**Comment by ajnyga** [comment]
After thinking about you comment about notifying reviewers if the review type is changed, I realized that currently the review type is not visible for the reviewer at all.

Should we maybe add some visual cue to the review workflow that shows the current review type? This would be visible in all steps. This way the change of the review type would make more sense to the reviewer.

**Comment by NateWr** [comment]
> Should we maybe add some visual cue to the review workflow that shows the current review type?

Yeah, I noticed this a while ago but it was tacked on to a low-priority issue: https://github.com/pkp/pkp-lib/issues/1488. It'd be nice to get that improved.

**Comment by ajnyga** [comment]
Great thanks! I will check that issue in detail and see what could be resolved here.

**Comment by ajnyga** [comment]
Hi @NateWr and @asmecher 
I did some initial work with adding the discussions grid to the reviewer workflow. The branch I am working with is here: https://github.com/ajnyga/pkp-lib/tree/open-review-enhancements-2976

With the grid permissions, I am using the same logic that the authors have.

However, there is one very big issue here which needs to be resolved.

While trying to get the grid visible in the reviewer workflow, I got stuck with permissions (which are extremely hard to understand at some times). I tracked down the issue to this function: https://github.com/pkp/pkp-lib/blob/ojs-stable-3_1_0/classes/security/authorization/internal/UserAccessibleWorkflowStageRequiredPolicy.inc.php#L79

I can of course add a case there for reviewers as well, but that is of course not enough, because reviewers are not regarded as stage participants in OJS.

So the question is, should we:
- Add reviewers as stage participants
- Just make the discussions grid work by extending _getAccessibleStageRoles and probably add similar exceptions to other places as well.

The first solution should probably mean that reviewers would not be visible in the participants grid (filter them out). They would be visible for the editors in the add a new discussion feature and also for authors in the case of open reviews. With this enabled, we could remove the "email reviewer" feature. Not sure if I am seeing all the implications of having the reviewers as stage participants, @asmecher?

The second solution would keep the reviewers separate from the stage participants as it is now. But it would possibly require some suspicious solutions.

What do you think?

**Comment by NateWr** [comment]
@asmecher has a much better grasp on the auth stuff than me so he'll need to decide which approach makes sense.

I can see treating reviewers as stage participants makes some sense. And it would fit nicely with plans to start giving other stage participants assignments with due dates like reviewers. We might be able to generalise the concept of assignments a bit more easily if reviewers are just one kind of stage participant.

**Comment by ajnyga** [comment]
Yes. As I said above I would probably be fairly easy to filter out the reviewers from the participants grid, if having them there is a problem (thinking of situations with several review rounds). And I am worried that the secondary solution will lead to a lot more tweaks and more code to maintain.

**Comment by ajnyga** [comment]
Hi @asmecher, I am continuing my work with this later today. I bet you are busy with many other issues as well, but I would need your comments to the questions above. Meanwhile I will start the work with adding the reviewers grid to the author workflow.

**Comment by asmecher** [comment]
Hmm, this is an old can of worms and I'm trying to recall the decision-making.

Currently we have three separate "views" on the workflow, with different entry points: the "author dashboard", for authors; the "reviewer" view; and the "workflow" view, for everyone else (assistants, editors, etc).

Of these three views, the reviewer view is the only one that *doesn't* use stage assignments. This reflects the very different way that reviewers are assigned (@NateWr, this is relevant to the upcoming addition of `vue.js` prettiness to the reviewer assignment process). Reviewer assignments are currently *only* in the `review_assignments` table, *not* `stage_assignments` (and the normal role structure). IMO these data models are different enough that it might not be a natural fit to add `stage_assignment` concerns to the reviewer assignments, unless you've got any big ideas for workflow reconciliation, @NateWr?

I suspect adding `stage_assignments` to reviewers would involve a really broad search for unintended consequences -- e.g. accidentally allowing reviewer access to the workflow tools, metadata views, etc. for double-blind reviews. Plus issues of data syncing. So I think it's a risk without a lot of upside.

I would suggest building smarter rules into the few areas you *do* want to potentially grant reviewer access to. @ajnyga, is there an example grid you want me to flesh this out a little for?

Open to other ideas, of course!

**Comment by ajnyga** [comment]
Hi @asmecher and @NateWr 

Thanks. Yes I agree that adding to the stage_assignments table could create unwanted results. Also the problem there are the multiple review rounds. 

The places where we need to overcome this are probably just:
1. Showing the grid
2. Listing the possible discussion participants
3. Showing the files grid in the query form

For 1, right now I am thinking of doing a custom grid handler for reviewer discussions which would of course be based on the default queries grid. I would just change the authorization methods used there and use the review_assignments for authorization. 

For 2, changes are needed here https://github.com/pkp/pkp-lib/blob/master/controllers/grid/queries/form/QueryForm.inc.php#L197. It needs to check a) that the current stage is review and b) the review type of each assignment before adding the reviewer and if #2717 is considered now, then the listing needs to be dependent on whether it is the author or the editorial staff doing the request.

So important question here: do you want me to tackle https://github.com/pkp/pkp-lib/issues/2717 right away as well? This would just be a diffe

For 3, I did not check the code yet, but I do not think it will be a problem. Probably it has the same stage_assignment check.

If I am going to the wrong direction, then just let me know.

**Comment by NateWr** [comment]
> IMO these data models are different enough that it might not be a natural fit to add stage_assignment concerns to the reviewer assignments, unless you've got any big ideas for workflow reconciliation, @NateWr?

The only thing I can think of is the introduction of workflow tasks with due dates (ie - copyedit these files, due by X). This may make stage assignments a bit more similar to review assignments, but wouldn't require they be merged. So yeah just keep as-is I guess.

**Comment by ajnyga** [comment]
I will keep them separate and when it is closer being finished we can evaluate it again. I have some more additions coming tomorrow and will try to include the conditional discussion grid authentication as well. I am including #2717 as well.

**Comment by asmecher** [comment]
@ajnyga, your plan makes sense to me -- but I do worry about reviewers being confused about having both the review grid and the discussions tool. Is it worth considering *just* giving them access to the reviews area, and excluding them from the discussion tool? We might need to clarify the role of the discussion tool as post-review, i.e. an internal discussion to the journal/press.

Alternately, how could we lead reviewers into having access to both, but keeping their comments etc. in the right place?

**Comment by asmecher** [comment]
(One way to avoid confusion could be for the discussions tool only to show up when a reviewer has been assigned to a discussion? Or when the review is closed?)

**Comment by ajnyga** [comment]
Hi @asmecher,

Once the discussions grid is displaying in the reviewer workflow without problems (almost there, I am maybe through the security policy jungle now) it should be fairly easy to place it anywhere we find suitable. 

I agree that adding more features to the reviewer workflow is risky because we need it to be as clear as possible. I think that it is easier to discuss this once the grid is there, I mean when the future is ready.

**Comment by asmecher** [comment]
Sure, we can run some UI/UX tests on it.

**Comment by ajnyga** [comment]
Hi @asmecher (and @NateWr if you are still interested in sticking around)

The discussions grid is almost ready now.

The problem with the grid was https://github.com/pkp/pkp-lib/blob/ojs-stable-3_1_0/classes/security/authorization/WorkflowStageAccessPolicy.inc.php which gets called from several places.

I first tried to create a reviewer specific grid handler and actually got quite far with that solution. However, this meant a lot of new code to maintain and there were still some stage assigment related access policies that were triggered. Those were very hard to debug.

So I came up with another solution. 

Originally I was thinking of just editing WorkflowStageAccessPolicy but soon realized that it would affect a lot of other things as well. However, I ended up creating a Query specific WorkflowStageAccessPolicy which calls an extended version of UserAccessibleWorkflowStageRequiredPolicy.inc.php which has a custom  _getAccessibleStageRoles function. This function adds reviewers to the stage if they have an active review assignment and is only used in the context of queries leaving the rest of the stage assignment policies intact.

I do not know if this is a unconventional way of handling this, but as you can see from the pull request, it was fairly straightforward way of doing it.

Here is a working version (together with a couple of additions to the reviewer grid and edit action):
https://github.com/pkp/pkp-lib/pull/3077

What is missing is the way to control the visible participants when the review is not open. In particular the situation where the current user is a reviewer and the review is not open. In these cases the author name should not of course be visible. It should happen around here https://github.com/ajnyga/pkp-lib/blob/a42d17efc47c62d0cb31f0b790070f3c8649451c/controllers/grid/queries/form/QueryForm.inc.php#L202 but I could not find an easy way to check whether the call is coming from the reviewer workflow. Any suggestions?

Another, more minor thing is that the review_assignments do not show which userGroupId the assignment belongs to. This could be relevant when the context has more than one type of reviewer roles? In my case it would be usefule here: https://github.com/ajnyga/pkp-lib/blob/a42d17efc47c62d0cb31f0b790070f3c8649451c/controllers/grid/queries/form/QueryForm.inc.php#L223

**Comment by ajnyga** [comment]
Note that I added a "login as" action to the reviewer grid, mentioned here: https://github.com/pkp/pkp-lib/issues/2508. Sorry, could not resist because while testing I was using it a lot.

**Comment by ajnyga** [comment]
With participant list in the query form I am now using the following logic:
- In open reviews: all users see eachother, including the reviewers, the see eachother (what do you think?) 
- in blind reviews:  authors only sees the editors; reviewers see the editors; editors see everybody
- in mixed reviews (both blind and open): authors only see editors on open reviewers: blind reviewers only see editors; open reviewers see authors and editors but not the blind reviewers

The logic is here: https://github.com/ajnyga/pkp-lib/blob/66046d346e192e67e205d5d54515f73ed852fec7/controllers/grid/queries/form/QueryForm.inc.php#L211-L250

I am happy to make to it more simple, just could not figure how :)

**Comment by ajnyga** [comment]
I added the reviewer grid to the author dashboard. It only shows when open reviews exist.

@NateWr while testing I noticed that in the editor workflow the "Thank reviewer" linkaction is not working. I think that my changes here do not affect that so is there a bug in 3.1 there?

**Comment by NateWr** [comment]
Maybe.

**Comment by ajnyga** [comment]
@NateWr I only have the notifications left here. The idea was to let the reviewer know if the review type is changed. However, it would probably make sense to have a similar notification for all the changes made after the request is sent and confirmed, meaning also the changes in the dates and the available files? What do you think?

**Comment by NateWr** [comment]
Yeah that makes sense.

**Comment by ajnyga** [comment]
It will now check whether review details are changed upon editing and if they are, a notification is send to the reviewer. The notification system is somewhat complex, so I am not sure whether this is the way to go.

**Comment by NateWr** [comment]
I think that's good. It is hard to know what the right kind of notification is, and we don't have a good overview of notification behaviour, so firing off an email is good enough for now, I think.

**Comment by ajnyga** [comment]
Hi @asmecher 
I think that you could check the code now: https://github.com/pkp/pkp-lib/pull/3077

I noticed that Nate committed a week ago some code that conflicts with the way I am extending the query participant list in the review stage. I will have to look through the changes he made. But the pr will basically work with 3.1.0.1 stable.

**Comment by NateWr** [comment]
@ajnyga Just saw your codepen about the icons. What do you think about [this](https://codepen.io/natewr/pen/gvNNqm)?

**Comment by ajnyga** [comment]
Good idea, if of course it seems that the icons add value to the UI. I was not too sure about that, just something that came to mind. What do you think about using the term non-blind (referring to the twitter conversation)?

This has been in stand still. Basically I have everything ready for 3.1.0.1 and almost ready version for the master branch as well but have been waiting for a conference call with folks at PKP. I have ask about that again and I think that there a couple of details that Alec has not commented yet.

**Comment by NateWr** [comment]
> What do you think about using the term non-blind (referring to the twitter conversation)?

Which twitter conversation?

**Comment by ajnyga** [comment]
https://twitter.com/ajnyga/status/968823360703606787

**Comment by NateWr** [comment]
I'm not a fan of "non-blind" because it's as ambiguous as "open" and doesn't clarify much. We could search for a word like "Identified", "Named", "Disclosed", "Visible", etc. A fully public review could be "Public".

This is the kind of thing where getting more input from practitioners is really valuable.

**Comment by ajnyga** [comment]
I like "Disclosed", but I think that native speakers should decide.

We could of course think "open review" as an umbrella term and "Public review" and "Disclosed review" could be under it.

I agree that more input is needed. The problem is that "open review" means very different things to different people. But I also think that at some point OJS will have tools for "public review" and we can not just use the term "open review" for everything when that happens.

(And of course the type of review where anyone can participate is Peer Review 2.0 :P)

**Comment by ajnyga** [comment]
Remaining features planned in a new issue here https://github.com/pkp/pkp-lib/issues/3738

**Comment by asmecher** [comment]
Thanks @ajnyga, this was huge.

**Comment by prechelt** [comment]
@NateWr commented on 2018-03-06:
> I'm not a fan of "non-blind" because it's as ambiguous as "open" and doesn't clarify much. We could search for a word like "Identified", "Named", "Disclosed", "Visible", etc. A fully public review could be "Public".

There is no way a single word can reliably convey all the different aspects involved. For full flexibility, I think one must name them individually, e.g. `AuthorsKnowReviewers`, `AuthorsKnowReviews`, `ReviewersKnowReviewers`, `ReviewersKnowReviews`, `ReviewersKnowAuthors`, `PublicKnowsReviewers`, `PublicKnowsReviews`, and so on. 

The front part is the actor involved, the back part is an aspect that could be known: actors, review text.
A classical **single-blind** review process becomes: `AuthorsKnowReviews`, `ReviewersKnowAuthors`, and perhaps (but sometimes not) `ReviewersKnowReviews`. 
And **double-blind** is `AuthorsKnowReviews`, and perhaps (but sometimes not) `ReviewersKnowReviews`. 

Cumbersome, yes. But clear. And versatile for many reviewing models:

**Triple-blind** models could include editors as ignorant actors. 
**Open** models exist in many variants -- that's why you need these aspect-wise definitions for them. 
**Truly open** models would include `Public` as an actor or `All` as aspect. 
`Public` is the same as "all actors", which shortens the descriptions of **fully truly open** reviewing to `PublicKnowsAll`.

Certainly not fun to display in a UI, but eventually the only solution, I guess, once open reviewing starts to (slowly) spread.

(in case you wonder: I am the person working on a [reviewqualitycollector](https://reviewqualitycollector.org) plugin.)

**Comment by marcbria** [comment]
Not sure where to post this.
Just found that NISO keep a nice summary of peer review terminology that could be useful in this discussion:
- Summary: https://groups.niso.org/higherlogic/ws/public/download/28900/Z39.106-2023-Peer_Review_Terminology.pdf
- https://www.niso.org/publications/z39106-2023-peerreview
- https://groups.niso.org/higherlogic/ws/public/download/28908/Peer_review_terminology.pdf
